broadcaster/_backends/pulsar.py

```diff
@@ -1,77 +1,119 @@
             self._client = await anyio.to_thread.run_sync(
                 lambda: pulsar.Client(self._service_url)
             )
-            self._producer = await anyio.to_thread.run_sync(
-                lambda: self._client.create_producer("broadcast")
-            )
-            self._consumer = await anyio.to_thread.run_sync(
-                lambda: self._client.subscribe(
-                    "broadcast",
-                    subscription_name="broadcast_subscription",
-                    consumer_type=pulsar.ConsumerType.Shared,
-                )
-            )
-            logging.info("Successfully connected to brokers")
+            logger.info("Successfully connected to Pulsar brokers")
         except Exception as e:
-            logging.error(e)
+            logger.error(f"Error connecting to Pulsar: {e}")
             raise e
 
     async def disconnect(self) -> None:
-        if self._producer:
-            await anyio.to_thread.run_sync(self._producer.close)
-        if self._consumer:
-            await anyio.to_thread.run_sync(self._consumer.close)
+        # Cancel all receiver tasks
+        for task in self._receiver_tasks.values():
+            task.cancel()
+        
+        # Wait for all receiver tasks to complete
+        await asyncio.gather(*self._receiver_tasks.values(), return_exceptions=True)
+        
+        # Prepare coroutines for closing producers and consumers
+        close_coros = [
+            anyio.to_thread.run_sync(producer.close)
+            for producer in self._producers.values()
+        ] + [
+            anyio.to_thread.run_sync(consumer.close)
+            for consumer in self._consumers.values()
+        ]
+        
+        # Add client close coroutine if client exists
         if self._client:
-            await anyio.to_thread.run_sync(self._client.close)
+            close_coros.append(anyio.to_thread.run_sync(self._client.close))
+        
+        # Execute all close operations concurrently
+        await asyncio.gather(*close_coros, return_exceptions=True)
+        
+        # Clear all containers
+        self._producers.clear()
+        self._consumers.clear()
+        self._receiver_tasks.clear()
+        self._client = None
+        
+        logger.info("Disconnected from Pulsar")
 
     async def subscribe(self, channel: str) -> None:
-        # In this implementation, we're using a single topic 'broadcast'
-        # So we don't need to do anything here
-        pass
+        if channel not in self._consumers:
+            consumer = await anyio.to_thread.run_sync(
+                lambda: self._client.subscribe(
+                    channel,
+                    subscription_name=f"broadcast_subscription_{channel}",
+                    consumer_type=pulsar.ConsumerType.Shared,
+                )
+            )
+            self._consumers[channel] = consumer
+            self._receiver_tasks[channel] = asyncio.create_task(self._receiver(channel, consumer))
```

Comment:
There is probably a need to do a quick `await asyncio.sleep(0)` here to allow the event loop to briefly begin the new coroutine.